### PR TITLE
Repository and UseCase

### PR DESCRIPTION
**Pull request by SandroBoka**
Repository and UseCase layers added.

Repository calls WeatherService and maps the WeatherResponse to WeatherModel which only has the weather information that is being used.

GetWeatherUseCase has been added so that the ViewModels don't have to call through the repository to get weather data.

### CODE REVIEW

File: WeatherApp/WeatherApp/UseCases/GetWeather.swift [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import Foundation
+
+protocol GetWeatherUseCaseProtocol {
+    func getWeather(cityName: String, completion: @escaping (Result<WeatherModel, ClientError>) -> Void)
+}
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/UseCases/GetWeather.swift]: newlines before and after

File: WeatherApp/WeatherApp/Repo/WeatherRepository.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        return WeatherModel(
+            temp: response.main.temp,
+            feelsLike: response.main.feelsLike,
+            description: weatherDescription,
+            humidity: response.main.humidity,
+            speed: response.wind.speed,
+            sunrise: response.system.sunrise,
+            sunset: response.system.sunset
+        )
+    }
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/Repo/WeatherRepository.swift]: create a custom init on the `WeatherModel` that takes `CurrentWeatherResponse` as parameter.

You can put it in a private extension in this file

File: WeatherApp/WeatherApp/Repo/WeatherRepository.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+protocol WeatherRepositoryProtocol {
+
+    func fetchWeather(for cityName: String, completion: @escaping (Result<WeatherModel, ClientError>) -> Void)
+
+}
+
+class WeatherRepository: WeatherRepositoryProtocol {
+
+    let weatherService: WeatherServiceProtocol
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/Repo/WeatherRepository.swift]: private

File: WeatherApp/WeatherApp/Repo/WeatherRepository.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    init(weatherService: WeatherServiceProtocol) {
+        self.weatherService = weatherService
+    }
+
+    func fetchWeather(
+        for cityName: String,
+        completion: @escaping (Result<WeatherModel, ClientError>) -> Void
+    ) {
+        weatherService.fetchWeather(for: cityName) { result in
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/Repo/WeatherRepository.swift:21]: memory leak, weak self needed

File: WeatherApp/WeatherApp/Repo/WeatherModel.swift [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+struct WeatherModel {
+
+    let temp: Double
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/Repo/WeatherModel.swift]: temperature

File: WeatherApp/WeatherApp/CityScreen/CityScreenViewModel.swift [hunk 6]
```diff
@@ -3,25 +3,25 @@ import SwiftUI
...
     private let router: RouterProtocol
-    private let weatherService: WeatherServiceProtocol
+    private let getWeatherUseCase: GetWeatherUseCaseProtocol
 
     @Published var city: String
-    @Published var weather: CurrentWeatherResponse?
+    @Published var weather: WeatherModel?
 
-    init(router: RouterProtocol, service: WeatherServiceProtocol, city: String) {
+    init(router: RouterProtocol, useCase: GetWeatherUseCaseProtocol, city: String) {
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityScreen/CityScreenViewModel.swift]: i would still like this to be names `getWeatherUseCase` to make it clearer what use case is in use

File: WeatherApp/WeatherApp/CityScreen/CityScreenViewModel.swift [hunk 7]
```diff
@@ -3,25 +3,25 @@ import SwiftUI
 class CityScreenViewModel: ObservableObject {
 
     private let router: RouterProtocol
-    private let weatherService: WeatherServiceProtocol
+    private let getWeatherUseCase: GetWeatherUseCaseProtocol
 
     @Published var city: String
-    @Published var weather: CurrentWeatherResponse?
+    @Published var weather: WeatherModel?
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityScreen/CityScreenViewModel.swift]: private(set) both published variables

File: WeatherApp/WeatherApp/CityListScreen/CityListView.swift [hunk 8]
```diff
@@ -37,5 +37,9 @@ struct CityListView: View {
     CityListView(
         viewModel: CityListViewModel(
             router: Router(navigationController: UINavigationController()),
-            service: WeatherService(client: NetworkClient())))
+            useCase: GetWeatherUseCase(
+                weatherRepo: WeatherRepository(weatherService: WeatherService(client: NetworkClient()))
+            )
+        )
+    )
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListView.swift]: Inline this with the last item

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift [hunk 9]
```diff
@@ -13,24 +13,24 @@ class CityListViewModel: ObservableObject {
         City(name: "Split")]
 
     private let router: RouterProtocol
-    private let weatherService: WeatherServiceProtocol
+    private let getWeatherUseCase: GetWeatherUseCaseProtocol
 
-    init(router: RouterProtocol, service: WeatherServiceProtocol) {
+    init(router: RouterProtocol, useCase: GetWeatherUseCaseProtocol) {
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift]: `useCase` -> `getWeatherUseCase`

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift [hunk 10]
```diff
@@ -13,24 +13,24 @@ class CityListViewModel: ObservableObject {
...
         self.router = router
-        self.weatherService = service
+        self.getWeatherUseCase = useCase
 
         fetchWeatherForAllCities()
     }
 
     func fetchTemperature(for city: City) {
-        weatherService.fetchWeather(for: city.name) { [ weak self ] result in
+        getWeatherUseCase.getWeather(cityName: city.name) { [ weak self ] result in
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift]: [ weak self ] -> [weak self]

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift [hunk 11]
```diff
@@ -13,24 +13,24 @@ class CityListViewModel: ObservableObject {
...
             guard let self = self else { return }
 
             switch result {
-            case .success(let weatherResponse):
+            case .success(let weatherModel):
                 if let index = self.cities.firstIndex(where: { $0.id == city.id }) {
-                    DispatchQueue.main.async { [ weak self ] in
-                        self?.cities[index].temperature = weatherResponse.main.temp
+                    DispatchQueue.main.sync { [ weak self ] in
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift]: [ weak self ] -> [weak self]

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift [hunk 12]
```diff
@@ -13,24 +13,24 @@ class CityListViewModel: ObservableObject {
...
+                        self?.cities[index].temperature = weatherModel.temp
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift]: use the at function from the #7

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift (hunk 11 above)
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift]: why sync here?

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift [hunk 13]
```diff
@@ -13,24 +13,24 @@ class CityListViewModel: ObservableObject {
         City(name: "Split")]
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift:13]: Published property `private(set)`

**Review by dnikolic98 — CHANGES_REQUESTED** [review]

File: WeatherApp/WeatherApp/CityListScreen/CityListView.swift (hunk 8 above)
Comment (SandroBoka) [reply WeatherApp/WeatherApp/CityListScreen/CityListView.swift]: Like it is now?

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift (hunk 11 above)
Comment (SandroBoka) [reply WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift]: It is now here anymore after the previous resolved conversation.

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift [hunk 14]
```diff
@@ -50,3 +51,9 @@ class CityListViewModel: ObservableObject {
     }
 
 }
+
+extension Array {
+    func at(_ index: Int) -> Element? {
+        return indices.contains(index) ? self[index] : nil
+    }
+}
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift]: once you merge the changes please delete this + `return` not needed

File: WeatherApp/WeatherApp/Repo/WeatherRepository.swift [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+        self.init(
+            temperature: response.main.temp,
+            feelsLike: response.main.feelsLike,
+            description: weatherDescription,
+            humidity: response.main.humidity,
+            speed: response.wind.speed,
+            sunrise: response.system.sunrise,
+            sunset: response.system.sunset
+        )
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/Repo/WeatherRepository.swift]: inlcine with last item

File: WeatherApp/WeatherApp/CityScreen/CityScreenView.swift [hunk 16]
```diff
@@ -48,8 +48,13 @@ struct CityScreenView: View {
...
-        city: "Atlantic City"))
+    CityScreenView(
+        viewModel: CityScreenViewModel(
+            router: Router(navigationController: UINavigationController()),
+            getWeatherUseCase: GetWeatherUseCase(
+                weatherRepo: WeatherRepository(weatherService: WeatherService(client: NetworkClient()))
+            ),
+            city: "Atlantic City"
+        )
+    )
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityScreen/CityScreenView.swift]: inline parenthesis

File: WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift [hunk 17]
```diff
@@ -13,24 +13,25 @@ class CityListViewModel: ObservableObject {
...
             guard let self = self else { return }
 
             switch result {
-            case .success(let weatherResponse):
-                if let index = self.cities.firstIndex(where: { $0.id == city.id }) {
-                    DispatchQueue.main.async { [ weak self ] in
-                        self?.cities[index].temperature = weatherResponse.main.temp
+            case .success(let weatherModel):
+                if let index = self.cities.firstIndex(where: { $0.id == city.id }),
+                   self.cities.at(index) != nil {
```
Comment (dnikolic98) [inline WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift:32]: guard let these (ignore if it was done on previous pr)

**Review by dnikolic98 — APPROVED** [review]

**Pull request merged**